protocols/imap/src/main/java/org/apache/james/imap/processor/base/UidMsnMapper.java

```diff
@@ -75,7 +79,41 @@ public synchronized void clear() {
     }
 
     public synchronized void addUid(MessageUid uid) {
-        this.addMapping(nextMsn(), uid);
+        if (isLastUid(uid)) {
+            addMapping(nextMsn(), uid);
+        } else {
+            addUidInMiddle(uid);
+        }
+    }
+
+    private boolean isLastUid(MessageUid uid) {
+        return msnToUid.isEmpty() ||
+            uid.asLong() > msnToUid.get(getLastMsn()).asLong();
```

Comment:
I prefer reading: 

```
uid.asLong() > msnToUid.get(getLastMsn()).asLong();
```

Than :

```
uid.compareTo(msnToUid.get(getLastMsn())) > 0;
```

Which involves a magic value, and knowing return codes of compareTo by heart....